### PR TITLE
Big changes

### PR DESCRIPTION
**Pull request by Moetez16**

### CODE REVIEW

File: server/gokbg3/grails-app/controllers/org/gokb/AdminController.groovy [hunk 1]
```diff
@@ -697,4 +697,21 @@ class AdminController {
     def result = [:]
     result
   }
+
+  def findTippDuplicates() {
+    log.debug("findTippDuplicates::${params}")
+    def result = [:]
+
+    params.pkg = Package.get(40893351)
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/controllers/org/gokb/AdminController.groovy:705]: Sure?

File: server/gokbg3/grails-app/controllers/org/gokb/AjaxSupportController.groovy [hunk 2]
```diff
@@ -1026,18 +1026,23 @@ class AjaxSupportController {
...
 
-                log.debug("Got ID: ${identifier_instance}")
-                // Link if not existing
-                if (!owner.ids.contains(identifier_instance)) {
-                  owner.ids.add(identifier_instance)
-                  owner.save()
-                }
-                else {
-                  flash.error = message(code:'identifier.link.unique')
+            if (!ident) {
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/controllers/org/gokb/AjaxSupportController.groovy]: else if?

File: server/gokbg3/grails-app/controllers/org/gokb/ComponentController.groovy [hunk 3]
```diff
@@ -82,39 +73,23 @@ class ComponentController {
...
-              AND (SELECT COUNT(c.combo_id) FROM combo AS c JOIN kbcomponent as kbc ON (c.combo_from_fk = kbc.kbc_id) WHERE
-                kbc.kbc_status_rv_fk <> :deleted
-                AND c.combo_to_fk = id.kbc_id
-                AND c.combo_type_rv_fk = :comboType
-                AND c.combo_status_rv_fk = :comboStatus
-                AND EXISTS (select kt.kbc_id from title_instance as kt where kt.kbc_id = c.combo_from_fk)) > 1
-              order by id.kbc_id limit :limit offset :offset ;
+            final String dquery = '''SELECT id.id_id FROM identifier_new AS id WHERE id.id_namespace_fk = :namespace
+              AND id.id_value in (select id_new.id_value FROM identifier_new AS id_new WHERE id_new.id_namespace_fk = :namespace group by id_new.id_value having count(id_new.id_value) > 1)
+              order by id.id_value limit :limit offset :offset ;
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/controllers/org/gokb/ComponentController.groovy:78]: sense of limit & offset?

File: server/gokbg3/grails-app/domain/org/gokb/cred/Identifier.groovy [hunk 4]
```diff
@@ -1,196 +1,184 @@
...
+  static belongsTo = [
+          tipp: TitleInstancePackagePlatform,
+          org: Org,
+          platform: Platform,
+          pkg: Package
   ]
 
+
   static constraints = {
+    uuid(nullable: true, unique: false, blank: false, maxSize: 2048)
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/Identifier.groovy]: UUIDs should never be nullable; nully entries should be remapped

File: server/gokbg3/grails-app/controllers/org/gokb/AdminController.groovy (hunk 1 above)
Comment (Moetez16) [reply server/gokbg3/grails-app/controllers/org/gokb/AdminController.groovy:705]: Only for Tests

File: server/gokbg3/grails-app/controllers/org/gokb/ComponentController.groovy (hunk 3 above)
Comment (Moetez16) [reply server/gokbg3/grails-app/controllers/org/gokb/ComponentController.groovy:78]: For overload

File: server/gokbg3/grails-app/domain/org/gokb/cred/Identifier.groovy (hunk 4 above)
Comment (Moetez16) [reply server/gokbg3/grails-app/domain/org/gokb/cred/Identifier.groovy]: I have to run a separate action for this. Using SQL to set the UUID is not that trivial

File: server/gokbg3/grails-app/domain/org/gokb/cred/KBComponent.groovy [hunk 5]
```diff
@@ -355,7 +350,7 @@ abstract class KBComponent implements Auditable{
       variantNames        : KBComponentVariantName,
       reviewRequests      : ReviewRequest,
       prices              : ComponentPrice,
-      languages            : KBComponentLanguage
+      languages           : KBComponentLanguage,
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/KBComponent.groovy]: the trailing , should raise a compilation error

File: server/gokbg3/grails-app/domain/org/gokb/cred/Org.groovy [hunk 6]
```diff
@@ -359,4 +360,12 @@ class Org extends KBComponent {
 
     result
   }
+
+  @Transient
+  String getIdentifierValue(idtype){
+    // As ids are combo controlled it should be enough just to call find here.
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/Org.groovy]: Is the comment still valid? IDs are not controlled by combo any more ...

File: server/gokbg3/grails-app/domain/org/gokb/cred/Package.groovy [hunk 7]
```diff
@@ -841,12 +723,22 @@ select tipp.id,
...
     }
   }
 
+  void findTippDuplicates() {
+
+  }
+
+  @Transient
+  String getIdentifierValue(idtype){
+    // As ids are combo controlled it should be enough just to call find here.
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/Package.groovy]: see above

File: server/gokbg3/grails-app/domain/org/gokb/cred/Platform.groovy [hunk 8]
```diff
@@ -465,4 +304,12 @@ class Platform extends KBComponent {
 
     result
   }
+
+  @Transient
+  String getIdentifierValue(idtype){
+    // As ids are combo controlled it should be enough just to call find here.
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/Platform.groovy]: see above

File: server/gokbg3/grails-app/domain/org/gokb/cred/TitleInstancePackagePlatform.groovy [hunk 9]
```diff
@@ -1624,78 +570,12 @@ class TitleInstancePackagePlatform extends KBComponent {
...
-        if(tippCombos.size() == 1) {
-          log.debug("tippMatchingByTitleID provider internal identifier matching by "+tippCombos.size() + ": "+ tippCombos.id)
-          return tippCombos[0]
-        }
-
-      }
-    }
-
+  String getIdentifierValue(idtype){
+    // As ids are combo controlled it should be enough just to call find here.
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/domain/org/gokb/cred/TitleInstancePackagePlatform.groovy]: see above

File: server/gokbg3/grails-app/i18n/messages_en.properties [hunk 10]
```diff
@@ -377,7 +377,9 @@ identifier.value.notNull=The Identifier must have a value!
 identifier.value.notUnique=An Identifier with this value already exists!
 identifier.value.noNamespace=No Namespace configured!
 identifier.create.error=Could not create Identifier!
-identifier.link.unique=This identifier is already connected to this component!
+identifier.no.unique.by.component=This identifier is already exists for this component!
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/i18n/messages_en.properties]: already exists or is already connected to

Comment (Moetez16) [reply server/gokbg3/grails-app/i18n/messages_en.properties]: is already exists

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 11]
```diff
@@ -675,7 +675,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'printIdentifier':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) in :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: printIdentifier, tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: printIdentifier, tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 12]
```diff
@@ -684,7 +684,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'onlineIdentifier':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) in :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: onlineIdentifier, tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: onlineIdentifier, tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 13]
```diff
@@ -699,7 +699,7 @@ class ExportService {
                                         titleIdNameSpace = platform.titleNamespace ? platform.titleNamespace.value : 'FAKE'
                                     }
 
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: titleIdNameSpace.toLowerCase(), tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: titleIdNameSpace.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 14]
```diff
@@ -708,7 +708,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'doiIdentifier':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: doiIdentifier.toLowerCase(), tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: doiIdentifier.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 15]
```diff
@@ -717,7 +717,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'zdb_id':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: zdbIdentifier.toLowerCase(), tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: zdbIdentifier.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 16]
```diff
@@ -726,7 +726,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'ezb_id':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :tippID and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: ezbIdentifier.toLowerCase(), tippID: tippID, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: ezbIdentifier.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 17]
```diff
@@ -735,7 +735,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'package_ezb_anchor':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :package and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: packageEzbAnchor.toLowerCase(), package: pkg.id, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: packageEzbAnchor.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/ExportService.groovy [hunk 18]
```diff
@@ -744,7 +744,7 @@ class ExportService {
                                     }
                                     break;
                                 case 'package_isci':
-                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i, Combo as c where LOWER(i.namespace.value) = :namespaceValue and c.fromComponent.id = :package and c.type = :kbTypeIDs and c.toComponent = i and c.status = :comboStatus', [namespaceValue: packageIsci.toLowerCase(), package: pkg.id, kbTypeIDs: RDStore.COMBO_TYPE_KB_IDS, comboStatus: RDStore.COMBO_STATUS_ACTIVE], [readOnly: true])
+                                    List identifiers = Identifier.executeQuery('select i.value from Identifier as i where LOWER(i.namespace.value) in :namespaceValue and i.tipp.id = :tippID', [namespaceValue: packageIsci.toLowerCase(), tippID: tippID], [readOnly: true])
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/ExportService.groovy]: convention: ... in (:namespaceValue)

File: server/gokbg3/grails-app/services/wekb/KbartImportService.groovy [hunk 19]
```diff
@@ -0,0 +1,1148 @@
...
+                    return null
+                    break;
+            }
+        }
+        else {
+            return null
+        }
+    }
+
+    static RefdataValue determineMediumRef(titleObj) {
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/services/wekb/KbartImportService.groovy:933]: Did we consider LAS:eR when extending the refdata category? How much sense does it make at all to restrict it to a controlled list?

File: server/gokbg3/grails-app/views/apptemplates/mainTemplates/_journal.gsp [hunk 20]
```diff
@@ -96,7 +96,7 @@
         <li><a href="#history" data-toggle="tab">Add to Title History</a></li>
       </g:if>
       <li><a href="#identifiers" data-toggle="tab">Identifiers <span
-          class="badge badge-warning">${d?.getCombosByPropertyNameAndStatus('ids', 'Active')?.size() ?: '0'}</span></a>
+          class="badge badge-warning">${d?.ids?.size() ?: '0'}</span></a>
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/views/apptemplates/mainTemplates/_journal.gsp]: inconsistency: elsewhere, the null checks are omitted

File: server/gokbg3/grails-app/views/apptemplates/mainTemplates/_title.gsp [hunk 21]
```diff
@@ -91,7 +91,7 @@
       <li><a href="#history" data-toggle="tab">Add to Title History</a></li>
     </g:if>
     <li><a href="#identifiers" data-toggle="tab">Identifiers <span
-        class="badge badge-warning">${d?.getCombosByPropertyNameAndStatus('ids', 'Active')?.size() ?: '0'}</span></a>
+        class="badge badge-warning">${d?.ids?.size() ?: '0'}</span></a>
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/views/apptemplates/mainTemplates/_title.gsp]: null checks are omitted at other places

File: server/gokbg3/grails-app/views/public/packageContent.gsp [hunk 22]
```diff
@@ -203,7 +203,7 @@
 
                 <li class="nav-item">
                     <a class="nav-link" href="#identifiers" data-toggle="tab" role="tab">Identifiers
-                        <span  class="badge badge-pill badge-info">${pkg?.getCombosByPropertyNameAndStatus('ids', 'Active').size()}</span>
+                        <span  class="badge badge-pill badge-info">${pkg?.ids.size()}</span>
```
Comment (croqueGrec09) [inline server/gokbg3/grails-app/views/public/packageContent.gsp]: May package be indeed null when viewing a package detail view?

File: server/gokbg3/src/main/groovy/org/gokb/CrossRefPkgRun.groovy [hunk 23]
```diff
@@ -185,7 +179,7 @@ class CrossRefPkgRun {
           reviewRequestService.raise(
             pkg,
             "TIPP rejected",
-            "TIPP ${json_tipp.name ?: json_tipp.title.name} coudn't be imported. ${(currentTippError as JSON).toString()}",
+            "TIPP ${json_tipp.name} coudn't be imported. ${(currentTippError as JSON).toString()}",
```
Comment (croqueGrec09) [inline server/gokbg3/src/main/groovy/org/gokb/CrossRefPkgRun.groovy]: couldn't ("l" missing)

**Review by croqueGrec09 — COMMENTED** [review]
remarks

**Review by croqueGrec09 — APPROVED** [review]
approved

**Pull request merged**